RedHatInsights/rhsm-subscriptions
# SWATCH-687 Perform nightly tallies via DB

## Detail
Perform the nightly tally for an org by querying the DB instead
of holding all hosts and their facts in memory to perform the
calculations.

Since everything we are tallying for an org has already been persisted
as 'buckets', we can perform our calculations via the query using
the bucket data. The query is using the cores and sockets on the bucket itself
because these fields are normailzed before being stored, unlike
the instance measurements.

An environment variable was added to allow enabling the old
legacy tally functionality should we identify any issues and would like
to disable the new tally logic.

This new tally logic should address the issue outlined in bug [SWATCH-49 Virtual satellites are not considered in tally](https://issues.redhat.com/browse/SWATCH-49) and will likely cause some discrepancy in IQE tests. Any virtual non-RHEL instance will now be properly tallied as VIRTUAL in both the DB based tally AND the legacy tally.

## References:
https://issues.redhat.com/browse/SWATCH-687

## How To Test
Use the steps below as a guide to testing, but please try and find ways to break this outside of what I describe. Using the insert-mock-hosts script is a good way to generate data for testing against a large number of hosts per org, BUT, it isn't the greatest at creating the various combos of data that we need to pull in from HBI. I'd suggest finding an org in stage and pulling a smallish subset of data from the inventory DB to test. Ideally the data should have some RHEL and non-RHEL hosts to tally.

**NOTES:**
* Be sure to replace `YOUR_TEST_ORG_ID`,  in the commands below, with the org_id that will have data in the inventory DB.
* Load some host records into the inventory DB for an org. This inventory data will be what we are basing our tally tests off of.

### Scenario 1: Existing Tally Data and re-tally occurs via DB Tally

Start with a fresh swatch DB, deploy the **main** branch.
```
dropdb -U rhsm-subscriptions rhsm-subscriptions && createdb -U rhsm-subscriptions rhsm-subscriptions  &&  DEV_MODE=true ./gradlew clean :bootRun
```

Run a nightly tally on the test org.
```
http :9000/hawtio/jolokia   type=exec   mbean='org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean'   operation='tallyOrg(java.lang.String)'   arguments:='["YOUR_TEST_ORG_ID"]'
```

Once complete, capture the key parts of the tally as follows.
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select s.product_id, m.measurement_type, m.uom, s.usage, s.sla, s.billing_provider, s.billing_account_id, s.granularity, m.value from tally_snapshots s, tally_measurements m where s.id = m.snapshot_id and s.org_id = 'YOUR_TEST_ORG_ID' order by s.product_id, s.granularity, m.measurement_type, m.uom, s.usage, s.sla, s.billing_provider, s.billing_account_id;" > main_tally

```

Deploy the **mstead/nightly-tally-via-db** branch (legacy tally NOT activated) and run a nightly tally for your test org.
```
DEV_MODE=true ./gradlew clean :bootRun
```
Run a nightly tally on the test org.
```
http :9000/hawtio/jolokia   type=exec   mbean='org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean'   operation='tallyOrg(java.lang.String)'   arguments:='["YOUR_TEST_ORG_ID"]'
```

Once complete, note in the logs that the tally was done via DB.
```
[INFO ] [org.candlepin.subscriptions.tally.InventoryAccountUsageCollector] - Running tally via DB for orgId=YOUR_TEST_ORG_ID
```

Capture the result of the tally as follows.
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select s.product_id, m.measurement_type, m.uom, s.usage, s.sla, s.billing_provider, s.billing_account_id, s.granularity, m.value from tally_snapshots s, tally_measurements m where s.id = m.snapshot_id and s.org_id = 'YOUR_TEST_ORG_ID' order by s.product_id, s.granularity, m.measurement_type, m.uom, s.usage, s.sla, s.billing_provider, s.billing_account_id;" > db_tally_on_existing_tally_data
```

Diff the SQL files you captured above. Any discrepancies you see should only be additional tallies for VIRTUAL non-rhel systems. This is because tallying by DB fixes https://issues.redhat.com/browse/SWATCH-49.

**NOTE**: The snip below shows net new tally records. The TOTAL value does not change for the product as we have always correctly updated the TOTALS correctly for VIRTUAL non RHEL.
```diff
$ diff main_tally db_tally_on_existing_tally_data
>  Satellite        | VIRTUAL          | CORES     |       |         | _ANY             | _ANY               | DAILY       |     4
>  Satellite        | VIRTUAL          | CORES     |       | _ANY    | _ANY             | _ANY               | DAILY       |     4
>  Satellite        | VIRTUAL          | CORES     | _ANY  |         | _ANY             | _ANY               | DAILY       |     4
>  Satellite        | VIRTUAL          | CORES     | _ANY  | _ANY    | _ANY             | _ANY               | DAILY       |     4
>  Satellite        | VIRTUAL          | INSTANCES |       |         | _ANY             | _ANY               | DAILY       |     1
>  Satellite        | VIRTUAL          | INSTANCES |       | _ANY    | _ANY             | _ANY               | DAILY       |     1
>  Satellite        | VIRTUAL          | INSTANCES | _ANY  |         | _ANY             | _ANY               | DAILY       |     1
>  Satellite        | VIRTUAL          | INSTANCES | _ANY  | _ANY    | _ANY             | _ANY               | DAILY       |     1
>  Satellite        | VIRTUAL          | SOCKETS   |       |         | _ANY             | _ANY               | DAILY       |     8
>  Satellite        | VIRTUAL          | SOCKETS   |       | _ANY    | _ANY             | _ANY               | DAILY       |     8
>  Satellite        | VIRTUAL          | SOCKETS   | _ANY  |         | _ANY             | _ANY               | DAILY       |     8
>  Satellite        | VIRTUAL          | SOCKETS   | _ANY  | _ANY    | _ANY             | _ANY               | DAILY       |     8
--- snip ---

```

### Scenario 2: Verify that tally results of initial tallies from both legacy and DB tallies are the same.
In this test, start from fresh and run a new tally using legacy tally and compare it to the DB tally from the last test. This PR includes a patch for fixing SWATCH-49 when legacy tally is run. This was done to ensure that both methods of tallying result in the same state.

Deploy the **mstead/nightly-tally-via-db** branch (legacy tally activated) and run a nightly tally for your test org.
```
dropdb -U rhsm-subscriptions rhsm-subscriptions && createdb -U rhsm-subscriptions rhsm-subscriptions  && LEGACY_NIGHTLY_TALLY_ENABLED=true DEV_MODE=true ./gradlew clean :bootRun
```
Run a nightly tally on the test org.
```
http :9000/hawtio/jolokia   type=exec   mbean='org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean'   operation='tallyOrg(java.lang.String)'   arguments:='["YOUR_TEST_ORG_ID"]'
```

Once complete, note in the logs that the legacy tally was run.
```
[INFO ] [org.candlepin.subscriptions.tally.InventoryAccountUsageCollector] - Running legacy nightly tally for orgId=YOUR_TEST_ORG_ID
```

Capture the result of the tally as follows.
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select s.product_id, m.measurement_type, m.uom, s.usage, s.sla, s.billing_provider, s.billing_account_id, s.granularity, m.value from tally_snapshots s, tally_measurements m where s.id = m.snapshot_id and s.org_id = 'YOUR_TEST_ORG_ID' order by s.product_id, s.granularity, m.measurement_type, m.uom, s.usage, s.sla, s.billing_provider, s.billing_account_id;" > legacy_tally_fresh_db
```

Diff the SQL files. There should be no differences.
```
$ diff db_tally_on_existing_tally_data legacy_tally_fresh_db
$
```
